test/bpf_tests/bpf_test.go

```diff
@@ -0,0 +1,470 @@
+	return func(t *testing.T) {
+		// create ctx with the max allowed size(4k - head room - tailroom)
+		ctx := make([]byte, 4096-256-320)
+
+		if progSet.setupProg != nil {
+			statusCode, result, err := progSet.setupProg.Test(ctx)
+			if err != nil {
+				t.Fatalf("error while running setup prog: %s", err)
+			}
+
+			ctx = make([]byte, len(result)+4)
+			nl.NativeEndian().PutUint32(ctx, statusCode)
+			copy(ctx[4:], result)
+		}
+
+		// Run test, input a
+		statusCode, _, err := progSet.checkProg.Test(ctx)
+		if err != nil {
+			t.Fatal("error while running check program:", err)
+		}
+
+		// Clear map value after each test
+		defer func() {
+			var key int32
+			value := make([]byte, resultMap.ValueSize())
+			resultMap.Lookup(&key, &value)
+			for i := 0; i < len(value); i++ {
+				value[i] = 0
+			}
+			resultMap.Update(&key, &value, ebpf.UpdateAny)
+		}()
+
+		var key int32
+		value := make([]byte, resultMap.ValueSize())
+		err = resultMap.Lookup(&key, &value)
+		if err != nil {
+			t.Fatal("error while getting suite result:", err)
+		}
+
+		// Detect the length of the result, since the proto.Unmarshal doesn't like trailing zeros.
+		valueLen := 0
+		valueC := value
+		for {
+			_, _, len := protowire.ConsumeField(valueC)
+			if len <= 0 {
+				break
+			}
+			valueLen += len
+			valueC = valueC[len:]
+		}
+
+		result := &SuiteResult{}
+		err = proto.Unmarshal(value[:valueLen], result)
+		if err != nil {
+			t.Fatal("error while unmarshalling suite result:", err)
+		}
+
+		for _, testResult := range result.Results {
+			// Remove the C-string, null-terminator.
+			name := strings.TrimSuffix(testResult.Name, "\x00")
+			t.Run(name, func(tt *testing.T) {
+				if len(testResult.TestLog) > 0 && testing.Verbose() || testResult.Status != SuiteResult_TestResult_PASS {
+					for _, log := range testResult.TestLog {
+						tt.Logf("%s", log.FmtString())
+					}
+				}
+
+				switch testResult.Status {
+				case SuiteResult_TestResult_ERROR:
+					tt.Fatal("Test failed due to unknown error in test framework")
+				case SuiteResult_TestResult_FAIL:
+					tt.Fail()
+				case SuiteResult_TestResult_SKIP:
+					tt.Skip()
+				}
+			})
+		}
+
+		if len(result.SuiteLog) > 0 && testing.Verbose() ||
+			SuiteResult_TestResult_TestStatus(statusCode) != SuiteResult_TestResult_PASS {
+			for _, log := range result.SuiteLog {
+				t.Logf("%s", log.FmtString())
+			}
+		}
+
+		switch SuiteResult_TestResult_TestStatus(statusCode) {
+		case SuiteResult_TestResult_ERROR:
+			t.Fatal("Test failed due to unknown error in test framework")
+		case SuiteResult_TestResult_FAIL:
+			t.Fail()
+		case SuiteResult_TestResult_SKIP:
+			t.SkipNow()
+		}
+	}
+}
+
+type suiteTestResult struct {
+	name string
+	logs []testLog
+	code byte
+}
+
+type testLog struct {
+	fmt  string
+	args []uint64
+}
+
+// A simplified version of fmt.Printf logic, the meaning of % specifiers changed to match the kernels printk specifiers.
```

Comment:
Could this contain some examples? It's not very clear what this does. Why is it called `FmtString()`?

Reply:
I added an example and references, I hope that makes it clearer. The reason for calling it `FmtString()` is because `String()` is already taken. I define a function on the `Log` struct which is defined by the protobuf generated code, which itself implements `String()` by default, but that of course doesn't do the string formatting magic we want.